--- a/ecm_classifier_default.c
+++ b/ecm_classifier_default.c
@@ -821,3 +821,2 @@ static struct ctl_table ecm_classifier_d
-	},
-	{ }
+	}
 };

--- a/ecm_db/ecm_db.c
+++ b/ecm_db/ecm_db.c
@@ -673,2 +673 @@ static struct ctl_table ecm_db_ctl_table
-	{ }
 };

--- a/ecm_interface.c
+++ b/ecm_interface.c
@@ -9601,3 +9601,2 @@ static struct ctl_table ecm_interface_ta
-	},
-	{ }
+	}
 };

--- a/frontends/sfe/ecm_sfe_common.c
+++ b/frontends/sfe/ecm_sfe_common.c
@@ -530,3 +530,2 @@ static struct ctl_table ecm_sfe_sysctl_t
-	},
-	{}
+	}
 };

--- a/ecm_classifier_mark.c
+++ b/ecm_classifier_mark.c
@@ -675,8 +675,7 @@ static struct ctl_table ecm_classifier_m
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_classifier_mark_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_classifier_dscp.c
+++ b/ecm_classifier_dscp.c
@@ -944,8 +944,7 @@ static struct ctl_table ecm_classifier_d
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_classifier_dscp_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_classifier_emesh.c
+++ b/ecm_classifier_emesh.c
@@ -4376,8 +4376,7 @@ static struct ctl_table ecm_classifier_e
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_classifier_sawf_emesh_udp_ipsec_port_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/frontends/ecm_front_end_ipv4.c
+++ b/frontends/ecm_front_end_ipv4.c
@@ -151,8 +151,7 @@ static struct ctl_table ecm_front_end_ip
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_front_end_ipv4_stop_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/frontends/sfe/ecm_sfe_ipv4.c
+++ b/frontends/sfe/ecm_sfe_ipv4.c
@@ -1148,2 +1148 @@ static struct ctl_table ecm_sfe_ipv4_ctl
-	{ }
 };
--- a/frontends/sfe/ecm_sfe_ipv6.c
+++ b/frontends/sfe/ecm_sfe_ipv6.c
@@ -1143,7 +1143,6 @@ static struct ctl_table ecm_sfe_ipv6_ctl
 		.proc_handler	= &ecm_sfe_ipv6_reject_acceleration_for_ipsec_handler,
 	},
 #endif
-	{ }
 };
 
 /*
--- a/frontends/sfe/ecm_sfe_multicast_ipv4.c
+++ b/frontends/sfe/ecm_sfe_multicast_ipv4.c
@@ -3226,8 +3226,7 @@ static struct ctl_table ecm_sfe_multicas
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_front_end_ipv4_mc_stop_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/frontends/sfe/ecm_sfe_multicast_ipv6.c
+++ b/frontends/sfe/ecm_sfe_multicast_ipv6.c
@@ -3219,8 +3219,7 @@ static struct ctl_table ecm_sfe_multicas
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_front_end_ipv6_mc_stop_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/frontends/ecm_front_end_ipv6.c
+++ b/frontends/ecm_front_end_ipv6.c
@@ -152,8 +152,7 @@ static struct ctl_table ecm_front_end_ip
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_front_end_ipv6_stop_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/frontends/cmn/ecm_bond_notifier.c
+++ b/frontends/cmn/ecm_bond_notifier.c
@@ -269,8 +269,7 @@ static struct ctl_table ecm_bond_notifie
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_bond_notifier_stop_handler,
-	},
-	{ }
+	}
 };
 
 void ecm_bond_notifier_stop(int num)
--- a/frontends/ppe/ecm_ppe_ipv6.c
+++ b/frontends/ppe/ecm_ppe_ipv6.c
@@ -1055,8 +1055,7 @@ static struct ctl_table ecm_ppe_ipv6_ctl
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_ppe_ipv6_vlan_passthrough_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_classifier_mscs.c
+++ b/ecm_classifier_mscs.c
@@ -1420,7 +1420,6 @@ static struct ctl_table ecm_classifier_m
 		.proc_handler	= &ecm_classifier_mscs_scs_udp_ipsec_port_handler,
 	},
 #endif
-	{ }
 };
 
 /*
--- a/ecm_classifier_pcc.c
+++ b/ecm_classifier_pcc.c
@@ -1366,8 +1366,7 @@ static struct ctl_table ecm_classifier_p
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_classifier_pcc_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_classifier_ovs.c
+++ b/ecm_classifier_ovs.c
@@ -2230,8 +2230,7 @@ static struct ctl_table ecm_classifier_o
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_classifier_ovs_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_state.c
+++ b/ecm_state.c
@@ -971,8 +971,7 @@ static struct ctl_table ecm_state_ctl_ta
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_state_file_output_mask_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_conntrack_notifier.c
+++ b/ecm_conntrack_notifier.c
@@ -449,8 +449,7 @@ static struct ctl_table ecm_conntrack_no
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_conntrack_notifier_stop_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/frontends/ppe/ecm_ppe_ipv4.c
+++ b/frontends/ppe/ecm_ppe_ipv4.c
@@ -997,8 +997,7 @@ static struct ctl_table ecm_ppe_ipv4_ctl
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_ppe_ipv4_vlan_passthrough_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_db/ecm_db_connection.c
+++ b/ecm_db/ecm_db_connection.c
@@ -5270,8 +5270,7 @@ static struct ctl_table ecm_db_connectio
 		.maxlen		= sizeof(int),
 		.mode		= 0666,
 		.proc_handler	= &ecm_db_connection_defunct_by_5tuple_mask_handler,
-	},
-	{ }
+	}
 };
 
 /*
--- a/ecm_tracker_udp.c
+++ b/ecm_tracker_udp.c
@@ -1219,6 +1219,5 @@ static struct ctl_table ecm_tracker_udp_
 		.mode		= 0644,
 		.proc_handler	= &ecm_tracker_udp_rtp_payload_types_handler,
-	},
-	{}
+	}
 };
 
--- a/frontends/ecm_front_end_common.c
+++ b/frontends/ecm_front_end_common.c
@@ -1254,6 +1254,5 @@ static struct ctl_table ecm_front_end_sy
 		.mode		= 0644,
 		.proc_handler	= &ecm_front_end_tcp_denied_ports_handler,
-	},
-	{}
+	}
 };
 
--- a/ecm_sdx/ecm_sdx.c
+++ b/ecm_sdx/ecm_sdx.c
@@ -217,6 +217,5 @@ static struct ctl_table ecm_sdx_table[]
 		.mode		= 0666,
 		.proc_handler	= &ecm_sdx_lan_prefix_handler,
-	},
-	{}
+	}
 };
 
--- a/examples/ecm_ae_select.c
+++ b/examples/ecm_ae_select.c
@@ -700,6 +700,5 @@ static struct ctl_table ecm_ae_select_sy
 		.mode		= 0644,
 		.proc_handler	= &ecm_ae_select_test_rule_handler,
-	},
-	{ }
+	}
 };
 
--- a/examples/ecm_sdx_pcc.c
+++ b/examples/ecm_sdx_pcc.c
@@ -935,6 +935,5 @@ static struct ctl_table ecm_sdx_pcc_sysc
 		.mode			= 0644,
 		.proc_handler		= &ecm_sdx_pcc_unregister_handler,
-	},
-	{ }
+	}
 };
 
--- a/examples/ecm_pcc_test.c
+++ b/examples/ecm_pcc_test.c
@@ -1558,6 +1558,5 @@ static struct ctl_table ecm_pcc_test_sys
 		.mode		= 0644,
 		.proc_handler	= &ecm_pcc_test_pcc_mirror_en_handler,
-	},
-	{ }
+	}
 };
 
--- a/examples/ecm_sfe_l2.c
+++ b/examples/ecm_sfe_l2.c
@@ -1210,6 +1210,5 @@ static struct ctl_table ecm_sfe_l2_sysct
 		.mode		= 0644,
 		.proc_handler	= &ecm_sfe_l2_defunct_by_port_handler,
-	},
-	{ }
+	}
 };
 
--- a/ecm_classifier_wifi.c
+++ b/ecm_classifier_wifi.c
@@ -721,8 +721,7 @@ static struct ctl_table ecm_classifier_w
 		.maxlen		= sizeof(int),
 		.mode		= 0644,
 		.proc_handler	= &ecm_classifier_wifi_enable_handler,
-	},
-	{ }
+	}
 };
 
 /*

--- a/ecm_conntrack_notifier.c
+++ b/ecm_conntrack_notifier.c
@@ -496,5 +496,5 @@ int ecm_conntrack_notifier_init(struct d
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
-	result = nf_conntrack_register_notifier(&init_net, &ecm_conntrack_notifier);
+	result = nf_conntrack_register_chain_notifier(&init_net, &ecm_conntrack_notifier);
 	if (result < 0) {
 		DEBUG_ERROR("Can't register nf notifier hook.\n");
 		goto init_cleanup_1;
@@ -504,6 +504,10 @@ int ecm_conntrack_notifier_init(struct d
 	}
 #else
-	nf_conntrack_register_notifier(&init_net, &ecm_conntrack_notifier);
+	result = nf_conntrack_register_notifier(&init_net, &ecm_conntrack_notifier);
+	if (result < 0) {
+		DEBUG_ERROR("Can't register nf notifier hook.\n");
+		goto init_cleanup_1;
+	}
 #endif
 
 	/*
@@ -524,7 +524,7 @@ int ecm_conntrack_notifier_init(struct d
 
 init_cleanup_2:
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
-	nf_conntrack_unregister_notifier(&init_net, &ecm_conntrack_notifier);
+	nf_conntrack_unregister_chain_notifier(&init_net, &ecm_conntrack_notifier);
 #else
 	nf_conntrack_unregister_notifier(&init_net);
 #endif
@@ -552,7 +552,7 @@ void ecm_conntrack_notifier_exit(void)
 	nf_ct_netns_put(&init_net, NFPROTO_IPV6);
 #endif
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
-	nf_conntrack_unregister_notifier(&init_net, &ecm_conntrack_notifier);
+	nf_conntrack_unregister_chain_notifier(&init_net, &ecm_conntrack_notifier);
 #else
 	nf_conntrack_unregister_notifier(&init_net);
 #endif
