--- a/include/linux/if_bridge.h
+++ b/include/linux/if_bridge.h
@@ -179,6 +179,12 @@ int br_multicast_list_adjacent(struct net_device *dev,
+void br_dev_update_stats(struct net_device *dev,
+			 struct rtnl_link_stats64 *nlstats);
+
 struct net_device *br_fdb_find_port(const struct net_device *br_dev,
 				    const unsigned char *addr,
 				    __u16 vid);
+struct net_device *br_fdb_find_port_rcu(const struct net_device *br_dev,
+					const unsigned char *addr,
+					__u16 vid);
 void br_fdb_clear_offload(const struct net_device *dev, u16 vid);
 bool br_port_flag_is_set(const struct net_device *dev, unsigned long flag);
 u8 br_port_get_stp_state(const struct net_device *dev);
@@ -188,11 +194,23 @@ u8 br_port_get_stp_state(const struct net_device *dev);
+static inline void br_dev_update_stats(struct net_device *dev,
+				       struct rtnl_link_stats64 *nlstats)
+{
+}
+
 static inline struct net_device *
 br_fdb_find_port(const struct net_device *br_dev,
 		 const unsigned char *addr,
 		 __u16 vid)
 {
 	return NULL;
 }
 
+static inline struct net_device *
+br_fdb_find_port_rcu(const struct net_device *br_dev,
+			 const unsigned char *addr, __u16 vid)
+{
+	return NULL;
+}
+
 static inline void br_fdb_clear_offload(const struct net_device *dev, u16 vid)
 {
 }
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -1012,6 +1012,9 @@ struct sk_buff {
 	__u8			csum_not_inet:1;
 #endif
 	__u8			unreadable:1;
+#ifdef CONFIG_SHORTCUT_FE
+	__u8			fast_forwarded:1;
+#endif
 	__u8			tc_depth:2;
 #if defined(CONFIG_NET_SCHED) || defined(CONFIG_NET_XGRESS)
 	__u16			tc_index;	/* traffic control index */
--- a/net/bridge/br_fdb.c
+++ b/net/bridge/br_fdb.c
@@ -239,6 +239,29 @@ static struct net_bridge_fdb_entry *br_fdb_find(struct net_bridge *br,
 	return fdb;
 }
 
+struct net_device *br_fdb_find_port_rcu(const struct net_device *br_dev,
+					const unsigned char *addr,
+					__u16 vid)
+{
+	const struct net_bridge_port *dst;
+	struct net_bridge_fdb_entry *f;
+	struct net_bridge *br;
+
+	lockdep_assert_in_rcu_read_lock();
+
+	if (!netif_is_bridge_master(br_dev))
+		return NULL;
+
+	br = netdev_priv(br_dev);
+	f = fdb_find_rcu(&br->fdb_hash_tbl, addr, vid);
+	if (!f)
+		return NULL;
+
+	dst = READ_ONCE(f->dst);
+	return dst ? dst->dev : NULL;
+}
+EXPORT_SYMBOL_GPL(br_fdb_find_port_rcu);
+
 struct net_device *br_fdb_find_port(const struct net_device *br_dev,
 				    const unsigned char *addr,
 				    __u16 vid)
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -521,6 +521,9 @@ config FAILOVER
 	  migration of VMs with direct attached VFs by failing over to the
 	  paravirtual datapath when the VF is unplugged.
 
+config SHORTCUT_FE
+	bool "Enables kernel network stack path for Shortcut  Forwarding Engine"
+
 config ETHTOOL_NETLINK
 	bool "Netlink interface for ethtool"
 	select DIMLIB
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -765,6 +765,29 @@ void br_port_flags_change(struct net_bri
 		br_recalculate_neigh_suppress_enabled(br);
 }
 
+void br_dev_update_stats(struct net_device *dev,
+			 struct rtnl_link_stats64 *nlstats)
+{
+
+	struct pcpu_sw_netstats *stats;
+
+	/* Is this a bridge? */
+	if (!(dev->priv_flags & IFF_EBRIDGE))
+		return;
+
+
+	stats = get_cpu_ptr(dev->tstats);
+
+	u64_stats_update_begin(&stats->syncp);
+	u64_stats_add(&stats->rx_packets, nlstats->rx_packets);
+	u64_stats_add(&stats->rx_bytes, nlstats->rx_bytes);
+	u64_stats_add(&stats->tx_packets, nlstats->tx_packets);
+	u64_stats_add(&stats->tx_bytes, nlstats->tx_bytes);
+	u64_stats_update_end(&stats->syncp);
+	put_cpu_ptr(stats);
+}
+EXPORT_SYMBOL_GPL(br_dev_update_stats);
+
 bool br_port_flag_is_set(const struct net_device *dev, unsigned long flag)
 {
 	struct net_bridge_port *p;
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3684,14 +3684,22 @@ static int xmit_one(struct sk_buff *skb,
 {
 	unsigned int len;
 	int rc;
-
+#ifdef CONFIG_SHORTCUT_FE
+	/* If this skb has been fast forwarded then we don't want it to
+	 * go to any taps (by definition we're trying to bypass them).
+	 */
+	if (!skb->fast_forwarded) {
+#endif
 	if (dev_nit_active(dev))
 		dev_queue_xmit_nit(skb, dev);
+#ifdef CONFIG_SHORTCUT_FE
+	}
+#endif
 
 #ifdef CONFIG_ETHERNET_PACKET_MANGLE
 	if (dev->eth_mangle_tx && !(skb = dev->eth_mangle_tx(dev, skb)))
 		return NETDEV_TX_OK;
 #endif
 
 	len = skb->len;
 	trace_net_dev_start_xmit(skb, dev);
@@ -5548,6 +5556,11 @@ void netdev_rx_handler_unregister(struct
 }
 EXPORT_SYMBOL_GPL(netdev_rx_handler_unregister);
 
+#ifdef CONFIG_SHORTCUT_FE
+int (*athrs_fast_nat_recv)(struct sk_buff *skb) __rcu __read_mostly;
+EXPORT_SYMBOL_GPL(athrs_fast_nat_recv);
+#endif
+
 /*
  * Limit the use of PFMEMALLOC reserves to those protocols that implement
  * the special handling of PFMEMALLOC skbs.
@@ -5596,6 +5609,10 @@ static int __netif_receive_skb_core(stru
 	int ret = NET_RX_DROP;
 	__be16 type;
 
+#ifdef CONFIG_SHORTCUT_FE
+	int (*fast_recv)(struct sk_buff *skb);
+#endif
+
 	net_timestamp_check(!READ_ONCE(net_hotdata.tstamp_prequeue), skb);
 
 	trace_netif_receive_skb(skb);
@@ -5634,6 +5651,15 @@ another_round:
 			goto out;
 	}
 
+#ifdef CONFIG_SHORTCUT_FE
+	fast_recv = rcu_dereference(athrs_fast_nat_recv);
+	if (fast_recv) {
+		if (fast_recv(skb)) {
+			ret = NET_RX_SUCCESS;
+			goto out;
+		}
+	}
+#endif
 	if (skb_skip_tc_classify(skb))
 		goto skip_classify;
 
--- a/net/netfilter/nf_conntrack_ecache.c
+++ b/net/netfilter/nf_conntrack_ecache.c
@@ -143,12 +143,23 @@ static int __nf_conntrack_eventmask_repo
 	rcu_read_lock();
 
 	notify = rcu_dereference(net->ct.nf_conntrack_event_cb);
-	if (!notify) {
+#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
+	if (!notify && !rcu_dereference_raw(net->ct.nf_conntrack_chain.head))
+#else
+	if (!notify)
+#endif
+	{
 		rcu_read_unlock();
 		return 0;
 	}
-
+#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
+		ret = atomic_notifier_call_chain(&net->ct.nf_conntrack_chain,
+			events | missed, (void *)item);
+		if (notify)
+			ret = notify->ct_event(events | missed, item);
+#else
 	ret = notify->ct_event(events | missed, item);
+#endif
 	rcu_read_unlock();
 
 	if (likely(ret >= 0 && missed == 0))
@@ -170,28 +181,36 @@ int nf_conntrack_eventmask_report(unsigned int eventmask, struct nf_conn *ct,
 				  u32 portid, int report)
 {
 	struct nf_conntrack_ecache *e;
-	struct net *net = nf_ct_net(ct);
+	struct nf_ct_event item;
+	unsigned int missed;
+	int ret;
+
+	if (!nf_ct_is_confirmed(ct))
+		return 0;
 
 	e = nf_ct_ecache_find(ct);
 	if (e == NULL)
 		return 0;
 
-	if (nf_ct_is_confirmed(ct)) {
-		struct nf_ct_event item = {
-			.ct = ct,
-			.portid	= e->portid ? e->portid : portid,
-			.report = report
-		};
-		/* This is a resent of a destroy event? If so, skip missed */
-		unsigned long missed = e->portid ? 0 : e->missed;
+	memset(&item, 0, sizeof(item));
 
-		if (!((eventmask | missed) & e->ctmask))
-			return 0;
+	item.ct = ct;
+	item.portid = e->portid ? e->portid : portid;
+	item.report = report;
 
-		atomic_notifier_call_chain(&net->ct.nf_conntrack_chain, eventmask | missed, &item);
+	/* This is a resent of a destroy event? If so, skip missed */
+	missed = e->portid ? 0 : e->missed;
+
+	ret = __nf_conntrack_eventmask_report(e, eventmask, missed, &item);
+	if (unlikely(ret < 0 && (eventmask & (1 << IPCT_DESTROY)))) {
+		/* This is a destroy event that has been triggered by a process,
+		 * we store the PORTID to include it in the retransmission.
+		 */
+		if (e->portid == 0 && portid != 0)
+			e->portid = portid;
 	}
 
-	return 0;
+	return ret;
 }
 #else
 int nf_conntrack_eventmask_report(unsigned int events, struct nf_conn *ct,
@@ -239,40 +258,26 @@ void nf_ct_deliver_cached_events(struct nf_conn *ct)
 {
-	unsigned long events, missed;
 	struct nf_conntrack_ecache *e;
 	struct nf_ct_event item;
-	struct net *net = nf_ct_net(ct);
+	unsigned int events;
+
+	if (!nf_ct_is_confirmed(ct) || nf_ct_is_dying(ct))
+		return;
 
 	e = nf_ct_ecache_find(ct);
 	if (e == NULL)
 		return;
 
 	events = xchg(&e->cache, 0);
 
-	if (!nf_ct_is_confirmed(ct) || nf_ct_is_dying(ct) || !events)
-		return;
-
-	/* We make a copy of the missed event cache without taking
-	 * the lock, thus we may send missed events twice. However,
-	 * this does not harm and it happens very rarely. */
-	missed = e->missed;
-
-	if (!((events | missed) & e->ctmask))
-		return;
-
 	item.ct = ct;
 	item.portid = 0;
 	item.report = 0;
 
-	atomic_notifier_call_chain(&net->ct.nf_conntrack_chain,
-			events | missed,
-			&item);
-
-	if (likely(!missed))
-		return;
-
-	spin_lock_bh(&ct->lock);
-		e->missed &= ~missed;
-	spin_unlock_bh(&ct->lock);
+	/* We make a copy of the missed event cache without taking
+	 * the lock, thus we may send missed events twice. However,
+	 * this does not harm and it happens very rarely.
+	 */
+	__nf_conntrack_eventmask_report(e, events, e->missed, &item);
 }
 #else
 void nf_ct_deliver_cached_events(struct nf_conn *ct)
@@ -373,3 +373,17 @@ EXPORT_SYMBOL_GPL(nf_conntrack_unregister_notifier);
 EXPORT_SYMBOL_GPL(nf_conntrack_unregister_notifier);
+
+#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
+int nf_conntrack_register_chain_notifier(struct net *net, struct notifier_block *nb)
+{
+	return atomic_notifier_chain_register(&net->ct.nf_conntrack_chain, nb);
+}
+EXPORT_SYMBOL_GPL(nf_conntrack_register_chain_notifier);
+
+int nf_conntrack_unregister_chain_notifier(struct net *net, struct notifier_block *nb)
+{
+	return atomic_notifier_chain_unregister(&net->ct.nf_conntrack_chain, nb);
+}
+EXPORT_SYMBOL_GPL(nf_conntrack_unregister_chain_notifier);
+#endif
 
 void nf_conntrack_ecache_work(struct net *net, enum nf_ct_ecache_state state)
