--- a/feeds/packages/net/nginx/files-luci-support/60_nginx-luci-support
+++ b/feeds/packages/net/nginx/files-luci-support/60_nginx-luci-support
@@ -7,6 +7,7 @@ location /ubus {
         ubus_interpreter;
         ubus_socket_path /var/run/ubus/ubus.sock;
         ubus_parallel_req 2;
+        ubus_script_timeout 600;
 }
 EOT
 	fi
--- a/feeds/packages/net/nginx/files-luci-support/luci.locations
+++ b/feeds/packages/net/nginx/files-luci-support/luci.locations
@@ -1,14 +1,21 @@
+log_not_found off;
 location /cgi-bin/luci {
 		index  index.html;
 		include uwsgi_params;
 		uwsgi_param SERVER_ADDR $server_addr;
 		uwsgi_modifier1 9;
+ 		uwsgi_send_timeout 600;
+		uwsgi_connect_timeout 600;
+		uwsgi_read_timeout 600;
 		uwsgi_pass unix:////var/run/luci-webui.socket;
 }
 location ~ /cgi-bin/cgi-(backup|download|upload|exec) {
 		include uwsgi_params;
 		uwsgi_param SERVER_ADDR $server_addr;
 		uwsgi_modifier1 9;
+ 		uwsgi_send_timeout 600;
+		uwsgi_connect_timeout 600;
+		uwsgi_read_timeout 600;
 		uwsgi_pass unix:////var/run/luci-cgi_io.socket;
 }
 
